packages/opentelemetry-sdk-trace-base/src/MultiSpanProcessor.ts

```diff
@@ -47,7 +47,7 @@ export class MultiSpanProcessor implements SpanProcessor {
     });
   }
 
-  onStart(span: Span, context: Context): void {
+  onStart(span: WriteableSpan, context: Context): void {
```

Comment:
Discussed in the SIG about this being a breaking change. However, I tested it out and [TypeScript doesn't actually seem to care](https://www.typescriptlang.org/play?ts=4.6.2#code/JYOwLgpgTgZghgYwgAgOpWJOAjANhAZQAc4RkBvAWAChk7kQ4BbCALmQGcwMQBzAbhoBfGjQS44HDsmKlkwJkXwtw09Jgg58sslVr1GLZAF5kAIgAWEXLgD2ZwdRr1OtlgAUArnmAIAshBgFrYAJgAUAJQUItQxNKCQsIgoOu5QtkhStlAUzvS2IARgcFBgYRwkIOzqWHiElRHsAG62wCGOcdTiktJ+AJ6p6Zkc2fKKyhCqMpVpGRBZOXouBUUlZRWk7DqNyC1tufoudBsgAHQjHt64vgFBoZGOLjGdCAVcnERbM0PzIzmmIAgAHdkP1BnMFpEaBVTitiqUwuRDGxzFYbPYhBF+MgAPQ45AAeQA0gB+GiAkFg74Qv6RWGFeFlJHMFGWax2MyY7F45AAYWyUAgCDAuD6DFsYEJRKAA) in practice. I don't think it's sound in this regard.

Reply:
The caller here is SDK and for SDK it's not breaking.
For the consumer it's not breaking in usual usecases but it could break unusual usecases like accessing the constructor of the class and use it for anything.

So I think technically it's breaking but I doubt the breakage is relevant in real world usecases.

Reply:
Sorry, i was trying to say I think this change is fine because typescript doesn't care. See the linked TS playground. It _should_ care because the user's on_start would no longer be a valid subtype of SpanProcessor.on_start. Idk why it ignores this issue

Reply:
I think the pattern with abstract `onStart` methods is not typical for typescript as it was created to support callback style APIs usually used in Javascript.

`onStart` allows to specify a concrete as typescript verifies that the caller provides this concrete type. But in our case it's the callee which relies on the type but this is not verified by typescript.

For provide a callback style APIs it's the other way around.

see [here](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMoAc4mQbwFDKHIhwC2EAXMgM5hSgDmA3AUdRGACoCe6EAFGF6UadRgEoqYABbBqLAL548CADZxq1ZAEEMWZMFLpVEcuC17s+IsnQBXAEargCYmRQBeZACJvLG-ZOLjQcPHyCwlS09CAMksgycrisNoSJ1AB0QnzIXtkQ-qmEUBx2UNjphYRKKbaOzq5wltrowPzx0YzJRcWl5QmymflVyEoB9ABucJAJwrk+fng1eKCQsIgoAJKWAApQAPZImvtQ3UT7IKhgcFBgVPzUmCBUluK5AHzIE-vAACYjJQYcjWAHlLtdbvcLlcbmAAMIOe6PLAvJ5vDyfb5-dGYn7-JbKNQaLS6J57Q4QY6nQzGUwQczIbZkg5HagnM6EQINZDQiFgB5PVFYN44UY9IoAeglaCeBi0cFUqn2AHcIL9kA5uMR9mAdJZalzgoDgdAwTDbvxebCEUjBXq0R8vniRbUbOkMlbbgj5p74Q4RjVxsApjNfQiAPy2lH24WOrH4moIC60Gh2yzzEAQZUxkDtFhJkAppqyryZ7OkrDtQnJ3WPKhMrDk1ns0tZnNNylsqB5lQ15DFjNtisgDtUnt4KXIBAK5yxZDK4AyfvIX7QYNq2Y5JL7ADWeEeHvBsIFwpYB99-GLp7wxcP5v5yJA4hYt4vV6fLDwMDsIAQYGAFw8pcTwIiezwyrGOBKN+v7-oBcBmiBDhgVQw4ikoB7GrQppHha0JIc++7oBkWGgrh-IIcBWAIs+hCTqR0DLtOioOIgO7IDA7KrpMG75HK2q6ruN4YCREBAthUBmnylpUSANEviJDGSeRl6IdRDiEUAA) for a sample

So we do not break the build here but the way how we use typescript doesn't ensure that consumers get the type they expect.

Reply:
Ah I think I found the reason why TypeScript behaves this way in the docs regarding `strictFunctionTypes` option:

> [...the setting only applies to functions written in function syntax, not to those in method syntax](https://www.typescriptlang.org/tsconfig#:~:text=During%20development%20of%20this%20feature%2C%20we%20discovered%20a%20large%20number%20of%20inherently%20unsafe%20class%20hierarchies%2C%20including%20some%20in%20the%20DOM.%20Because%20of%20this%2C%20the%20setting%20only%20applies%20to%20functions%20written%20in%20function%20syntax%2C%20not%20to%20those%20in%20method%20syntax%3A)

I added a few lines to your example to demonstrate [here](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMoAc4mQbwFDKHIhwC2EAXMgM5hSgDmA3AUdRGACoCe6EAFGF6UadRgEoqYABbBqLAL548CADZxq1ZAEEMWZMFLpVEcuC17s+IsnQBXAEargCYmRQBeZACJvLG-ZOLjQcPHyCwlS09CAMksgycrisNoSJ1AB0QnzIXtkQ-qmEUBx2UNjphYRKKbaOzq5wltrowPzx0YzJRcWl5QmymflVyEoB9ABucJAJwrk+fng1eKCQsIgoAJKWAApQAPZImvtQ3UT7IKhgcFBgVPzUmCBUluK5AHzIE-vAACYjJQYcjWAHlLtdbvcLlcbmAAMIOe6PLAvJ5vDyfb5-dGYn7-JbKNQaLS6J57Q4QY6nQzGUwQczIbZkg5HagnM6EQINZDQiFgB5PVFYN44UY9IoAeglaCeBi0cFUqn2AHcIL9kA5uMR9mAdJZalzgoDgdAwTDbvxebCEUjBXq0R8vniRbUbOkMlbbgj5p74Q4RjVxsApjNfQiAPy2lH24WOrH4moIC60Gh2yzzEAQZUxkDtFhJkAppqyryZ7OkrDtQnJ3WPKhMrDk1ns0tZnNNylsqB5lQ15DFjNtisgDtUnt4KXIBAK5yxZDK4AyfvIX7QYNq2Y5JL7ADWeEeHvBsIFwpYB99-GLp7wxcP5v5yJA4hYt4vV6fLDwMDsIAQYGAFw8pcTwIiezwyrGOBKN+v7-oBcBmiBDhgVQw4ikoB7GrQppHha0JIc++7oBkWGgrh-IIcBWAIs+hCTqR0DLtOioOIgO7IDA7KrpMG75HK2q6ruN4YCREBAthUBmnylpUSANEviJDGSeRl6IdRDiEcok5hBAqAIPQ6C6r8+yUiAADkuoINIEAIOxnR-gAYj+f4ASAOlaJxpzeOQMj7Oq1DcOAcAAB7eCs4DQPASAQSOLKdicACyHDSH5qCBdcwUckB94oTF8TxooeB8bscVUk5sGuWlQWZV41jnORUbga8cZ4oVBYpjAACMQqxRSVJJb5vxVRl8yihej49ehoz5n2MAAEw9aOXYDSlQ3pSFo3ZdJE05lNCgzYWuowAAzItpVduVLnQutNW4Ftx47c1UHTb2h0cQALGdfUXc5cGXDdm3jXaaG4Ao01AA). Because SpanProcessor uses "method syntax" in the code, TS doesn't perform the strict check and this change should be fine, I think.